src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Snippets.java

```diff
@@ -0,0 +1,461 @@
+        public boolean isLastPagePartial() {
+            return (total % pageSize) != 0;
+        }
+
+        /**
+         * The number of pages.
+         */
+        public int pageCount() {
+            var pages = total / pageSize;
+            return isLastPagePartial() ? pages + 1 : pages;
+        }
+
+        /**
+         * The number of elements in the last page.
+         */
+        public int lastPageSize() {
+            if (total == 0) return 0;
+            var remaining = total % pageSize;
+            return remaining == 0 ? pageSize : remaining;
+        }
+    }
+
+    /**
+     * Generate bytecodes for loading a collection of elements, support using pagination to avoid
+     * overloading the 64k code limit.
+     */
+    public static abstract class CollectionSnippetBuilder {
+        /**
+         * Tested page size of string array
+         */
+        public static final int STRING_PAGE_SIZE = 8000;
+
+        /**
+         * Tested page size of enum array
+         */
+        public static final int ENUM_PAGE_SIZE = 5000;
+
+        /**
+         * Good enough for average ~30 bytes per element
+         */
+        public static final int DEFAULT_PAGE_SIZE = 2000;
+
+        /**
+         * Arbitary default values based on 15K code size on ~30 bytes per element
+         */
+        protected static final int DEFAULT_THRESHOLD = 512;
+
+        protected ClassDesc elementType;
+        protected ClassDesc ownerClassDesc;
+        protected ClassBuilder clb;
+
+        // Default values enable pagination by default
+        protected String methodNamePrefix = "csb" + Integer.toHexString(hashCode()) + "Page";
+        protected int activatePagingThreshold = DEFAULT_THRESHOLD;
+        protected int pageSize = DEFAULT_PAGE_SIZE;
+
+        /**
+         * @param elementType The element type
+         */
+        protected CollectionSnippetBuilder(ClassDesc elementType) {
+            this.elementType = Objects.requireNonNull(elementType);
+        }
+
+        /**
+         * Enable pagination if the count of elements is larger than the given threshold.
+         *
+         * @param methodNamePrefix The method name prefix for generated paging helper methods
+         * @param pageSize         The page size
+         * @param threshold        The element count to actiave the pagination
+         */
+        public CollectionSnippetBuilder enablePagination(String methodNamePrefix, int pageSize, int threshold) {
+            return this.pageSize(pageSize)
+                       .activatePagingThreshold(threshold)
+                       .methodNamePrefix(methodNamePrefix);
+        }
+
+        /**
+         * Enable pagination if the count of elements is larger than pageSize or DEFAULT_THRESHOLD
+         */
+        public CollectionSnippetBuilder enablePagination(String methodNamePrefix, int pageSize) {
+            return enablePagination(methodNamePrefix, pageSize, Math.min(pageSize, DEFAULT_THRESHOLD));
```

Comment:
shouldn't it use the current `activatePagingThreshold` value?  Otherwise, this method will override the threshold being set to a non-default value.

Same comment for `enablePagination(String methodNamePrefix)`

Reply:
I think about this, and it depends on how you look at the API.

The idea is that those three parameters are essential to enable pagination from disabled state. A default value is used when user is to enable pagination not specifying an explicit value, so a clear default is adapted rather than lingering with current setting.

Otherwise, user can use property methods to change page size or threshold.

I like the idea of use methodNamePrefix as the enable/disable check. However,  trying to explain the value of page size and threshold is current value, which would be default value without setting before is a little bit more complex, more so if trying to explain choosing the smaller value of page size and default threshold where we honestly have no idea if user have set a specific value or not.

I switched default to disable in the latest commit. And if it helps, I can remove those two enablePagination methods using default values.